### ISSUE TITLE
Add support for EXPLAIN PLAN FOR <query>

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#3594]
For instance:
- Oracle: EXPLAIN PLAN FOR >Your query<
- DB2: EXPLAIN PLAN SET queryno = xxx for >Your query<
- MS SQL Server: Set SHOWPLAN_ALL ON >Your query<
- Informix: SET EXPLAIN
- Sybase ASE: Set SHOWPLAN_ALL ON >Your query<

---

See also:
http://web.synametrics.com/top10performancetips.htm

**Comment by stangls** [comment]
I would love to have this feature to estimate the number of rows for huge queries.
I hope, based on this explain dsl syntax one could also implement a `dsl.estimate( org.jooq.Select selectStatement )`

Postgres: https://www.postgresql.org/docs/8.3/static/row-estimation-examples.html
H2: http://www.h2database.com/html/performance.html#explain_plan

**Comment by lukaseder** [comment]
Hmm interesting thought - a standardised total cardinality estimate might indeed be an interesting thing. It wouldn't just work for `Select`, but for any `Query` (e.g. estimate the number of updated rows in an `Update`). On the other hand, there can be "huge queries" that aggregate billions of rows into only a few dozen aggregated rows, so I'm not sure if it's really such a useful feature in general...

What's the use-case behind your idea?

Note, this particular issue is not yet well defined - including it is not yet known whether we should standardise the display of execution plans or just dump them directly to the result...

**Comment by stangls** [comment]
The use case is fairly simple.
We want to provide a paginated result for huge data-sets.

Every incoming request specifies a few filter conditions which are mapped via typesafe jooq queries to the db (possibly implying join operations). The intermediate result is always a jooq select statement. From this intermediate select we execute two queries:
1. estimate the un-paginated result (currently using very slow `count(*)`) and
2. request the paginated result (via offset & limit).

The paginated response then looks something like this (json):
```
{
  limit: 4
  offset: 20
  total: 27982
  elements [ "a", "b", "c", "d" ]
}
```

currently, `total` comes from either `count(*)` or is `null` for very long-running queries.
based on postgres, we could do stuff like this: https://wiki.postgresql.org/wiki/Count_estimate
but we also suppert H2 and other databases.

**Comment by stangls** [comment]
the estimate for select would be totally sufficient.
estimating update queries is not a benefit for us (yet?).

**Comment by lukaseder** [comment]
Interesting use-case. Well, it's going to be an estimate only, so I'm guessing that youre API / UI will be able to handle the error? I'm pretty sure that H2's estimates are terribly off for even simple queries (btw: are you using H2 just for development / testing?)

In any case, I've created a separate feature request for this: https://github.com/jOOQ/jOOQ/issues/6823 

You can implement it easily on your side by running a vendor-specific explain on the jOOQ query yourself. Just run

```java
DSL.using(configuration)
   .fetch("explain {0}", actualQuery);
```

**Comment by lukaseder** [comment]
@stangls I'm looking into this now. What would be your expected behaviour in the case of H2? The explain output seems pretty useless to me...

**Comment by lukaseder** [comment]
Another H2 example:

```sql
EXPLAIN ANALYZE SELECT * FROM (SELECT 1 UNION ALL SELECT 2)
```

Output:

```sql
SELECT
    _0."1"
FROM (
    (SELECT
        1
    FROM SYSTEM_RANGE(1, 1))
    UNION ALL
    (SELECT
        2
    FROM SYSTEM_RANGE(1, 1))
) _0
    /* (SELECT
        1
    FROM SYSTEM_RANGE(1, 1)
        /++ PUBLIC.RANGE_INDEX ++/
        /++ scanCount: 2 ++/)
    UNION ALL
    (SELECT
        2
    FROM SYSTEM_RANGE(1, 1)
        /++ PUBLIC.RANGE_INDEX ++/
        /++ scanCount: 2 ++/)
     */
    /* scanCount: 3 */
```

**Comment by lukaseder** [comment]
I've implemented this for some databases: #6823. I will close this issue as a duplicate, as ultimately, there's not really a difference between the issues.
